src/extra/components/pulses.py

```diff
@@ -169,7 +310,9 @@ def select_trains(self, trains):
 
         res = copy(self)
         res._source = self._source.select_trains(trains)
-        res._key = self._key.select_trains(trains)
+
+        if res._key is not None:
```

Comment:
When would `res._key` ever be `None`?

Reply:
Thanks, fixed!

Reply:
Got carried away here with "Thanks, fixed!". Proper reply would be, it used to be while I drafted the feature 🙃